### ISSUE TITLE
FFJORD: Calculating for multiple final timesteps

### ISSUE THREAD

**Issue by ishitamed19** [tensorflow/probability#972]
In reference to the FFJORD bijector, I can see in its argument final_time, it only supports a single time-step (i.e. scalar).

Now, what I want is to have it calculate for multiple time-steps (i.e. say given an initial_state=x0 and initial_time = 0.0 and final_times=[1.0,2.0,3.0], then it should return calculated values for these 3 time steps). For reference, the original implementation by the authors of FFJORD paper in Pytorch supports calculating for multiple time steps: https://github.com/rtqichen/ffjord

Is there a way to implement the same in Tensorflow?

TIA

**Comment by kochkov92** [comment]
Hi @ishitamed19, indeed computing intermediate values is fun!

I think all of the bijectors in TFP implement `_forward`, `_inverse` etc. methods with a single output, which allows them to be chained together and combined with distributions.

Here are a few ways to get the intermediate states with the current bijector API:
(1) Directly evolve the ode given by `state_derivative_fn` using an ODE solver (e.g. `tfp.math.ode.DormandPrince`) where you can provide `solution_times` to be any partitioning of the interval `[initial_time, final_time]`.

(2) "Split" one `FFJORD` bijector into N chains of `FFJORD` bijectors that share the same `state_derivative_fn`: (1st computes 0 -> `final_time / N`, 2nd chains the first + `final_time / N` -> `2 * final_time / N`, etc.)

**Comment by brianwa84** [comment]
@ishitamed19 can you say more about the use-case you're going for? If all you need is ODE integration, [`tfp.math.ode`](https://www.tensorflow.org/probability/api_docs/python/tfp/math/ode) could be a good place to start. If you want the trace-augmented solve, perhaps there are parts of FFJORD that we can pull out to some functions exposed under tfp.math.ode?

If you think of a FFJORD bijector as a bijective function mapping X->Y, then you can't really have more degrees of freedom in Y than you did in X.

That said, bijectors can support kwargs, e.g. for conditioning, see tfb.RealNVP, and perhaps it would be useful to have FFJORD return some additional intermediate solutions e.g. when a `intermediate_solution_times` kwarg is provided to `forward`? Or this particular bijector could even expose an additional public function to do such computations.

**Comment by brianwa84** [comment]
For an example of conditioning kwargs: https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/bijectors/real_nvp_test.py#L105-L135

**Comment by ishitamed19** [comment]
> Hi @ishitamed19, indeed computing intermediate values is fun!
> 
> I think all of the bijectors in TFP implement `_forward`, `_inverse` etc. methods with a single output, which allows them to be chained together and combined with distributions.
> 
> Here are a few ways to get the intermediate states with the current bijector API:
> (1) Directly evolve the ode given by `state_derivative_fn` using an ODE solver (e.g. `tfp.math.ode.DormandPrince`) where you can provide `solution_times` to be any partitioning of the interval `[initial_time, final_time]`.
> 
> (2) "Split" one `FFJORD` bijector into N chains of `FFJORD` bijectors that share the same `state_derivative_fn`: (1st computes 0 -> `final_time / N`, 2nd chains the first + `final_time / N` -> `2 * final_time / N`, etc.)

Hi @kochkov92 . Thanks for your reply! I'm already using the Dormand Prince. But as you can see in the `tfp.bijectors.FFJORD`, it passes only a scalar final_time to the ode solver's `solution_times`. i.e. `solution_times = [final_time]`.

Regarding (2), I too had thought of doing similar to that. A sample implementation is attached in the screenshot. However, I couldn't figure out how to proceed with training this (i.e should I iteratively do a forward pass with each `tfd.TransformedDistribution` for finding the loss?)

So what I'm trying to do is, I pick some `x_0 at time=t0` from the actual distribution, and then using the FFJORD, I want the predicted values at `times=[t1, t2, ... , tk]`, hence returning `[x1, x2, .., xk]`. So how can I calculate the gradients for this such that it also takes into account the `time`?

<img width="1053" alt="Screenshot 2020-06-18 at 12 36 04 AM" src="https://user-images.githubusercontent.com/6681634/84940010-d7fab380-b0fc-11ea-95ef-b45a7c636862.png">

**Comment by ishitamed19** [comment]
> @ishitamed19 can you say more about the use-case you're going for? If all you need is ODE integration, [`tfp.math.ode`](https://www.tensorflow.org/probability/api_docs/python/tfp/math/ode) could be a good place to start. If you want the trace-augmented solve, perhaps there are parts of FFJORD that we can pull out to some functions exposed under tfp.math.ode?
> 
> If you think of a FFJORD bijector as a bijective function mapping X->Y, then you can't really have more degrees of freedom in Y than you did in X.
> 
> That said, bijectors can support kwargs, e.g. for conditioning (see [realnvp](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/real_nvp_default_template) "condition_kwargs"), and perhaps it would be useful to have FFJORD return some additional intermediate solutions e.g. when a `intermediate_solution_times` kwarg is provided to `forward`? Or this particular bijector could even expose an additional public function to do such computations.

Thanks @brianwa84 for your reply! 
So, I need to use the FFJORD model (that too with the hutchinson estimator). What I'm trying to do is, I pick some `x_0 at time=t0` from the actual distribution, and then using the FFJORD, I want the predicted values at `times=[t1, t2, ... , tk]`, hence returning `[x1, x2, .., xk]`. 

Basically, `tfp.bijectors.FFJORD` currently is some function `f(x,t)`: where this `t` is a scalar (fixed using final_time). What I want is the ability to pass a 1-D tensor to `t` (`final_times=[t1, t2, .. ,tk]`). How can kwargs be used in this context? 

Also, we need to make sure that it not only returns the solution for those timesteps i.e. `[x1, x2, .. xk]`, but also the `log_prob` (for calculating the total loss) values for all these values,  as defined in `tfp.distributions.TransformedDistribution`.
